subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPlugin.java

```diff
@@ -287,7 +305,90 @@ private static void configureTestTaskOrdering(TaskContainer tasks) {
         tasks.withType(Test.class).configureEach(test -> test.shouldRunAfter(jarTasks));
     }
 
-    private static void configureBuiltInTest(Project project, JvmSoftwareComponentInternal component) {
+    // TODO: We should be configuring the main feature in the the sub-plugins, like application and java-library.
+    // Since the feature itself represents the software product being built, the sub-plugins, which decide which
+    // products are built, should be responsible for configuring the feature.
+    private static SingleTargetJvmFeature configureMainFeature(Project project, DefaultJvmSoftwareComponent component) {
+
+        ExtensionContainer extensions = project.getExtensions();
+        JavaPluginExtension javaExtension = extensions.findByType(JavaPluginExtension.class);
+
+        SourceSet mainSourceSet = javaExtension.getSourceSets().create(SourceSet.MAIN_SOURCE_SET_NAME);
+
+        // We construct the feature manually instead of with component.getFeatures().create() since this feature
+        // should have empty capabilities, representing the "implicit capability" as opposed to the default, which
+        // derives the capabilities from the feature name.
+        SingleTargetJvmFeature mainFeature = project.getObjects().newInstance(DefaultSingleTargetJvmFeature.class,
+            SourceSet.MAIN_SOURCE_SET_NAME, mainSourceSet, Collections.emptyList(),
+            project, ConfigurationRoles.INTENDED_CONSUMABLE, false
+        );
+        component.getFeatures().add(mainFeature);
+
+        // TODO: Should all features also have this variant? Why just the main feature?
```

Comment:
I don't think so.  A sourceless feature makes sense to me.

Reply:
Follow-up question: Should all features _with sources_ have this variant?

Reply:
Not every published library of a project with source publishes the source, even if the publishing setup is otherwise simple, so I think no.

Reply:
This variant does not get published, since it has a directory artifact, which is not published. This variant is currently used to support jacoco coverage reporting.

Reply:
It is separate from the similarly named sources variant which supports the actual sources jar. IMO they should be the same variant, since they provide essentially the same information.

Reply:
Ah, ok. What's the use case for a project needing to access another project's source directl locally like this?

Reply:
Inter-project? Probably none. Ideally future iterations of the variant API would allow us to declare "project local" variants which are only used internally to the project to support local features, like jacoco

Reply:
Updated the comment to explain that this is very similar to the other variant and we could consider consolidating them.